### DISCUSSION TITLE
Scoped Selectors For JS

### DISCUSSION THREAD

**Discussion by fweth** [withastro/roadmap#834] [Proposal]
### Body

# Summary

Add scoped JS selectors, working with `createNameHash` as used for the awesome scoped CSS selectors

# Background & Motivation

I think it's good if component's are self contained and can be moved between projects. If I query an element by ID (or something else) in a component, I always have to make sure that the ID is unique on my page, or, if I copy my component to a different project, that the ID is unique there as well. Also I think it would be comfortable to use simple pseudo-IDs which are only unique per component, like `btn-submit`.

# Goals

- Better isolated components.
- Making it more comfortable to work with vanilla JS, pushing the idea of not needing a JS framework for simple tasks on content driven websites (the kind which Astro targets so well).

# Example
Ideally you have the selector as a given function, like `Astro.querySelector`, `Astro.getElementById` or even `Astro.$`. Alternatively, if the hash of the current component could be accessed in JS (like `Astro.hash`), one could manually create such a selector. (But I think that currently only hashes are applied if scoped CSS is present, that could be a problem.)

**Comment by florian-lefebvre** [comment]
I think this could help you achieve this? https://github.com/lilnasy/gratelets/tree/main/packages/scope

**Reply by fweth** [reply]
Awesome, thanks! Since Astro already built a query transformer for more complex CSS rules (to add the hash), it would also be sweet to access this function in JS, additionally to the hash itself.
